pkg/controllers/user/pipeline/controller/pipelineexecution/pipelineexecution.go

```diff
@@ -129,3 +147,40 @@ func (l *Lifecycle) initLogs(obj *v3.PipelineExecution) error {
 	}
 	return nil
 }
+
+func (l *Lifecycle) getHeadCommit(execution *v3.PipelineExecution) (string, error) {
+	pipeline := execution.Spec.Pipeline
+	if !utils.ValidSourceCodeConfig(pipeline.Spec) {
+		return "", errors.New("invalid pipeline definition")
+	}
+	sourceCodeConfig := pipeline.Spec.Stages[0].Steps[0].SourceCodeConfig
```

Comment:
Just to confirm that there is no chance that len(stages) == 0 or len(steps) == 0

Reply:
It is ensured by the above ValidSourceCodeConfig() check